### ISSUE TITLE
Better syntax for ## and #=

### ISSUE THREAD

**Issue by chenglou** [facebook/reason#849]
cc @bobzhang

Currently, for BuckleScript (and hopefully jsoo) we have `foo##bar` for dot access, and `foo##bar #= 1` for object field assignment.

The feedback we've gotten is that people are scared of "hash hash" ¯\_(ツ)_/¯

So we can special-case hash hash and hash equal to something else in Reason. Proposal:

1. `foo..bar .= baz`
2. `foo->bar <- baz`
3. `foo."bar" = baz`

Third one might or might not work.

**Comment by bsansouci** [comment]
Yes I will confirm that `##` is too heavy for me https://github.com/bsansouci/tapknight/blob/6c3d7371a5c67209d75ee6bc309db755d8cf4081/src/client.re#L57, especially formatted like this ^.

If spacing was added as a special case, life would be a tiny bit nicer. 

Could we get `foo->bar = baz` to work? otherwise I'd prefer 2nd.

Getting sidetracked:
It'd be interesting if we could open a syntax locally to a scope. For example, it'd be cool to locally override `{}` and `foor.bar = bas` to be `[%bs.obj {}]` and `foo##bar #= baz`. That overriding could maybe look like
```
let obj = JS.(
  bar.baz = 10; /* bar##baz #= 10; */
  let a = {}; /* let a = [%bs.raw "{}"] */
 print_endline bar.baz; /* console.log(bar##baz); */
 bar
);

obj.baz /* error, this is of type Js.obj and not a record with a field bas */
```

It would be like you open a module. It would probably become a bit strange if you need to use records inside a `JS.(...)` block.

**Comment by chenglou** [comment]
> It'd be interesting if we could open a syntax locally to a scope 

Might be dangerous if you refactor code and pull e.g. a function out.

The thing about `foo->bar = baz` and `foo."bar" = baz` and `foo..bar = baz` to infer what to desugar `=` to, is that I'm not sure if this is possible: `myRefToFooBar = baz`. In which case this'd fail. Others know better.

**Comment by bobzhang** [comment]
I like proposal 1, another important thing is to make merlin easy to auto-complete
cc @let-def

**Comment by chenglou** [comment]
Adding a 4th proposal that @bobzhang mentioned.

**Comment by freebroccolo** [comment]
Out of these options I would prefer (3) or (4). I don't like (1) much because I feel like having more instances of `.` and `..` in the syntax is just going to look confusing. I already kind of wish we would only have `...`.

Regarding merlin, the prefix that gets sent to merlin is determined by the editor, so I don't think the choice of which syntax we use here will have much impact on that (the editor modes will need to be updated regardless).

One other possibility to keep in mind is that once we have modular implicits, we might be able to overload `#` so we wouldn't even need to have `##`. We might not even need to wait for that if we were willing to use an approach like `ppx_implicits` currently does.

**Comment by chenglou** [comment]
Ok, spiritually bumping the priority of this a bit.

The name mangling part of BS, `foo##bar_baz` (which compiles to `foo.bar`) is gonna trip a lot of people. Proposal 3 fixes this. We don't have to worry about the assignment operator for now.

Tldr `foo."bar_baz"` which desugars to ??? (cc @bobzhang) which compiles to the correct `foo.bar_baz`.

cc @bsansouci @SanderSpies.

**Comment by bobzhang** [comment]
see my thoughts here: https://github.com/bloomberg/bucklescript/issues/999

**Comment by iddan** [comment]
How about `obj['property'] = value`
